### ISSUE TITLE
How call a method/function at the end of every cycle of FSM

### ISSUE THREAD

**Issue by Eeliya** [GetmeUK/ContentTools#107]
I am using setInterval to call some custom method every X amount of time . as far as I understood the FSM does the same thing.

I think using interval is not a very good idea.

So I want to use FSM instead of setInterval. Is this possible or using interval is fine?

> My task: I want to keep track of changes in elements size and positions.

**Comment by anthonyjb** [comment]
Hi @Eeliya the FSM (Finite State Machine) is purely about state management, it allows you to say if the application is in this state then it's can only move to the following states. FSM is used by HTMLString to manage the state of the parser, e.g if the parser encounters a `<` it might move to a state of `OPENING_TAG` at which point it will only accept the next character in the range of `[a-z]` - I hope I've explained that well enough.

So to answer you question the FSM doesn't use `setInterval` - however I do use setInterval for monitoring the document and implementing undo/redo (e.g I take a regular snapshot of the editable document), and updating the status of tools to determine if they should be enabled, toggled, disabled, etc.

Whether this is a good idea or not is arguable, another common approach - certainly for undo/redo - is  to implement a message system where a stream of messages to change the document can be logged and run in reverse to provide undo support. This does mean that code needs to be put in place to revert as well as apply a change, but it's more efficient and less prone to issues. This approach works in content tools because we can snapshot the HTML and users current selection and expect the result of a revert/undo to that shapshot to be the same no matter what the state of the editor is.

I'm still on the fence about the decision I made here, however if you're calling `setInterval` to trigger a custom function periodically I don't see a problem with this so long as you ensure that the frequency of calls is not causing the browsers performance to suffer.
